Documentation: JS Getting started (#226)
SAP/cloud-sdk: docs/js/getting-started.md

#### TypeScript

- **`tsconfig.json`**: Configuration file for `TypeScript`. This is not needed in the plain `JavaScript` version.
- **`tslint.json`**: Configuration file for `tslint`.

#### Continuous Delivery

- **`Jenkinsfile`**: Jenkins pipeline definition file for quality assurance. It uses the [SAP Cloud SDK's Continuous Delivery Toolkit](https://github.com/SAP/cloud-s4-sdk-pipeline).
- **`pipeline_config.yml`**: Pipeline configuration file for the Jenkins pipeline.

#### Cloud Foundry

- **`manifest.yml`**: The deployment descriptor file for `Cloud Foundry in SAP Cloud Platform`.

#### Local development

- **`src/`**: Source code for the initial application.

#### SDK specific

- **`systems.json`+`credentials.json`**: Allows you to maintain destinations for testing purposes.
- **`sap-cloud-sdk-analytics.json`**: Only if you have agreed to usage analytics during the initialization of your project. You can find more information about anonymous usage analytics [in the CLI's repository](https://github.com/SAP/cloud-sdk-cli/blob/master/usage-analytics.md).

## Run the project

To run the application locally, simply run the following command:

```shell
npm run start:dev
```

> 

Reviewer: ```suggestion
This will start a local server in watch mode, so that subsequent changes will automatically trigger a restart of the server.
```

Go to `http://localhost:3000` and you should get a `Hello, World!` in response. Open `src/main.ts` and switch the port from `3000` to `8080`. The corresponding line should then look like this:

```ts
await app.listen(process.env.PORT || 8080);
```

Since `nest` was started in watch mode, it should detect this change and restart the server.

## Deploy the project on Cloud Foundry

> ### Installation

Reviewer: ```suggestion
### Prerequisites
```

> The `Cloud Foundry` CLI deploys your application to SAP Cloud Platform. You can find installation instructions for all common platforms in the [`Cloud Foundry documentation`](https://docs.cloudfoundry.org/cf-cli/install-go-cli.html). We recommend to use a package manager for that. If you are using `chocolatey` on Windows, please find the instructions [here](https://chocolatey.org/packages/cloudfoundry-cli).

Reviewer: ```suggestion
The Cloud Foundry CLI comes in handy when you want to deploy your application to SAP Cloud Platform. You can find installation instructions for all common platforms in the [`Cloud Foundry documentation`](https://docs.cloudfoundry.org/cf-cli/install-go-cli.html). We recommend to use a package manager for that. If you are using `chocolatey` on Windows, please find the instructions [here](https://chocolatey.org/packages/cloudfoundry-cli).
```

<> <BuildBadge /> { } <LicenseBadge /> </>

## Installation

> 

Author: Is this not good enough?

## CLI Usage

The CLI can initialize a nest-based project or (the more common case) add everything you need to develop for SAP Cloud Platform to an existing project no matter what backend framework you use.
If there are any incompatibilities, please let us know in the [issues](https://github.com/SAP/cloud-sdk-cli/issues/new/choose)!

> 

Reviewer: This is supposed to be the getting started section for the JS SDK. As a reader I now expect to be introduced in how to use the SDK, but the first sentence I read is - to install the CLI. Huh? SDK or CLI, what are we talking about?

I would rather expect to read something about the purpose of the SDK and that it can be used in applications on SAP Cloud Platform. To create such an application we provide a command line interface, that allows you to scaffold or enhance an application with the missing parts to use the SDK. And then we can say: 'To install the CLI...' because then the reader will know where we come from.

To install the CLI globally, run:

```shell
$ npm install -g @sap-cloud-sdk/cli
```

### Update

As long as the CLI version is less than `1.0.0`, run the following to update to the latest version. Please keep in mind, that these updates can have breaking changes as per the [semver spec](https://semver.org/#spec-item-4).

```shell
$ npm install -g @sap-cloud-sdk/cli@latest
```

## CLI Usage

The CLI can initialize a nest-based project or (the more common case) add everything you need to develop for SAP Cloud Platform to an existing project no matter what backend framework you use.
If there are any incompatibilities, please let us know in the [issues](https://github.com/SAP/cloud-sdk-cli/issues/new/choose)!

> ### Installation

Reviewer: This is the second installation section.

First run this command to install the CLI

```shell
npm install -g @sap-cloud-sdk/cli
```

### Usage

```shell
sap-cloud-sdk [COMMAND]
```
To get the CLI's version, run:
```shell
sap-cloud-sdk (-v|--version|version)
```
To get a list of all commands, run:
```shell
$ sap-cloud-sdk --help
```

### Create a new project

> To create a new project run the CLI's init command in the project folder.

Reviewer: ```suggestion
To create a new project run the CLI's `init` command in the project folder.
```

```shell
$ sap-cloud-sdk init
```

It will guide you through the initialization, create the necessary files and add necessary dependencies.
If you run it in an empty folder, it will ask if you want to initialize a project using [@nest/cli](https://github.com/nestjs/nest-cli).

> The CLI will already install all the necessary dependencies for the project, so this might take a minute. If everything worked correctly, you should see output like this:

Reviewer: ```suggestion
The CLI will already install all the necessary dependencies for the project, so this might take a minute. If everything worked correctly, you should see the following output:
```

```shell
+---------------------------------------------------------------+
| ✅ Init finished successfully.                                |
|                                                               |
| 🚀 Next steps:                                                |
| - Run the application locally (`npm run start:dev`)           |
| - Deploy your application (`npm run deploy`)                  |
|                                                               |
| 🔨 Consider setting up Jenkins to continuously build your app |
| Use `sap-cloud-sdk add-cx-server` to create the setup script  |
+---------------------------------------------------------------+
```

## Project files and folders
<!--not sure about this part-->

The project contains the following files and folders, among others, to get you started with the SAP Cloud SDK for JavaScript:

#### NPM / Project

> - **`package.json`**: Specifies dependencies, metadata and user-defined scripts. The application comes with some predefined scripts and dependencies, that will be explained in detail in the course of this group of tutorials.

Reviewer: tutorials?

- **`.npmrc`**: The **`npm`** configuration file. The SAP Cloud SDK consists of some generic libraries, that are available as Open Source Software and service libraries for the whitelisted SAP S/4HANA APIs, referred to as the Virtual Data Model (VDM). In the scaffolding we specify the registry for the `@sap` scope, where the VDM libraries are published.

#### TypeScript

- **`tsconfig.json`**: Configuration file for `TypeScript`. This is not needed in the plain `JavaScript` version.

> - **`tslint.json`**: Configuration file for `tslint`, the de facto default linter for `TypeScript`.

Reviewer: Unfortunately tslint is deprecated and we should update this as soon as possible ;)
```suggestion
- **`tslint.json`**: Configuration file for `tslint`.
```

#### Continuous Delivery

- **`Jenkinsfile`**: Jenkins pipeline definition file for quality assurance. It uses the [SAP Cloud SDK's Continuous Delivery Toolkit](https://github.com/SAP/cloud-s4-sdk-pipeline).
- **`pipeline_config.yml`**: Pipeline configuration file for the Jenkins pipeline.

#### Cloud Foundry

- **`manifest.yml`**: The deployment descriptor file for `Cloud Foundry in SAP Cloud Platform`.

#### Local development

- **`src/`**: Source code for the initial application.

#### SDK specific

- **`systems.json`+`credentials.json`**: Allows you to maintain destinations for testing purposes.
- **`sap-cloud-sdk-analytics.json`**: Only if you have agreed to usage analytics during the initialization of your project. You can find more information about anonymous usage analytics [in the CLI's repository](https://github.com/SAP/cloud-sdk-cli/blob/master/usage-analytics.md).

## Run the project

To run the application locally, simply run the following command:

```shell

> npm run start:dev

Reviewer: Maybe mention that this command starts the server in watch mode (this is mentioned later on, but `:dev` is the reason and this is not mentioned anywhere).

```

Go to `http://localhost:3000` and you should get a `Hello, World!` in response. Open `src/main.ts` and switch the port from `3000` to `8080`. The corresponding line should then look like this:

```ts
await app.listen(process.env.PORT || 8080);

> ```

Reviewer: We should explain why (because this will automatically be started using port 8080 in CF). But I think `process.env.PORT` already fixes this, so this step is not needed anymore
@florian-richter Can you confirm that?
@CharlesDuboisSAP Maybe you can double check by trying to push an app without changing this.

Since `nest` was started in watch mode, it should detect this change and restart the server.

## Deploy the project on Cloud Foundry

> ### Installation

Reviewer: Installation of what. We were in the context of deploying the project, why do I have to install something again? (Add an introduction)

> The `Cloud Foundry` CLI deploys your application to SAP Cloud Platform. You can find installation instructions for all common platforms in the [`Cloud Foundry documentation`](https://docs.cloudfoundry.org/cf-cli/install-go-cli.html). We recommend to use a `package manager` for that. If you are using `chocolatey` on Windows, please find the instructions [here](https://chocolatey.org/packages/cloudfoundry-cli).

Reviewer: ```suggestion
The `Cloud Foundry` CLI deploys your application to SAP Cloud Platform. You can find installation instructions for all common platforms in the [`Cloud Foundry documentation`](https://docs.cloudfoundry.org/cf-cli/install-go-cli.html). We recommend to use a package manager for that. If you are using `chocolatey` on Windows, please find the instructions [here](https://chocolatey.org/packages/cloudfoundry-cli).
```

### Login

> <!--Do i need to explain this part with more detail?-->

Reviewer: Not for now, I think 😊

:::note

> If you don't have a [SAP Cloud Platform](https://account.hana.ondemand.com/) account you need to create one.

Reviewer: ```suggestion
If you don't have an [SAP Cloud Platform](https://account.hana.ondemand.com/) account you need to create one.
```

- The command specified under the `command` attribute tells the `buildpack` what command to issue to start the application.

When everything works as expected, you should get output that looks something like this:

```shell
Waiting for app to start...

name:              <YOUR-APPLICATION-NAME>
requested state:   started
routes:            <YOUR-APPLICATION-NAME>.cfapps.eu10.hana.ondemand.com
last uploaded:     Thu 21 Mar 14:05:32 CET 2019
stack:             cflinuxfs3
buildpacks:        nodejs

type:            web
instances:       1/1
memory usage:    256M
start command:   node index.js
     state     since                  cpu    memory        disk           details
#0   running   2019-03-21T13:05:47Z   0.0%   16M of 256M   126.8M of 1G
```

The application will be running at the `routes` URL, you can also make sure that the application works correctly by running the start command, this command can be different than the one shown above.

Should the application not work for whatever reason, you can call the following command to access the logs:

```Shell
cf logs <YOUR-APPLICATION-NAME> --recent
```

## Additional features

> For productive use, your app should be bound to a server, and also implement user authentication and authorization.

Reviewer: ```suggestion
For productive use, your app should be bound to a server and implement user authentication and authorization.
```

Reviewer: I don't understand bound to a server. Can you explain?

Author: bound to a server -> linked to one or more databases?

    random-route: true
    services:
      - my-destination
```

#### XSUAA Service

Secondly, we need an instance of the `XSUAA service`. The `XSUAA service` is responsible for issuing access tokens that are necessary to talk to other services, like the destination service. For this service, we will need a bit of extra configuration in the form of a configuration file. Create a file called `xs-security.json` with the following content:

```JSON
{
  "xsappname": "<YOUR-APPLICATION-NAME>",
  "tenant-mode": "shared"
}
```

The value for `xsappname` again has to be unique across the whole of `Cloud Foundry in SAP Cloud Platform`, so make sure to choose a unique name or prefix.

Now, execute the following command:

```shell
cf create-service xsuaa application my-xsuaa -c xs-security.json
```

And, as before, add the newly created services to the services section of your `manifest.yml`.

The final `manifest.yml` should look like this:

```YAML
applications:
  - name: <YOUR-APPLICATION-NAME>
    path: deployment/
    buildpacks:
      - nodejs_buildpack
    memory: 256M
    command: node index.js
    random-route: true
    services:
      - my-destination
      - my-xsuaa
```

> Finally, we can replace the parameter of `execute` with an object whose key `destinationName` refers to the name of the destination we defined earlier. If you chose a different name than `Server`, make sure to use it here accordingly.

Reviewer: Maybe we should also use a different destination name than 'Server'. This is way too general. Maybe something like `S4_SYSTEM` or similar is more appropriate.
